File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Integer[] count_array = new Integer[128];
        int left = 0;
        int right = 0;
        int max_sub_len = 0;
        while(right < s.length()){
            char r = s.charAt(right);
            Integer index = count_array[r];
            if(index != null && index >= left && index < right) left = index + 1;

            max_sub_len = Math.max(max_sub_len, right- left + 1);

            count_array[r] = right;
            right++;
        }
        return max_sub_len;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l, r = 0, 0
        current_set = set()
        max_len = 0
        while r < len(s):
            if s[r] not in current_set:
                current_set.add(s[r])
                r += 1
                max_len = max(max_len, r - l)
            else:
                current_set.remove(s[l])
                l += 1
        return max_len

        
        